dm/worker/server_test.go

```diff
@@ -731,3 +732,35 @@ func loadSourceConfigWithoutPassword(c *C) *config.SourceConfig {
 	sourceCfg.From.Password = "" // no password set
 	return sourceCfg
 }
+
+func (t *testServer) TestServerDataRace(c *C) {
+	var (
+		masterAddr   = tempurl.Alloc()[len("http://"):]
+		keepAliveTTL = int64(1)
+	)
+	etcdDir := c.MkDir()
+	ETCD, err := createMockETCD(etcdDir, "http://"+masterAddr)
+	c.Assert(err, IsNil)
+	defer ETCD.Close()
+	cfg := NewConfig()
+	c.Assert(cfg.Parse([]string{"-config=./dm-worker.toml"}), IsNil)
+	cfg.Join = masterAddr
+	cfg.KeepAliveTTL = keepAliveTTL
+	cfg.RelayKeepAliveTTL = keepAliveTTL
+
+	s := NewServer(cfg)
+	defer s.Close()
+	go func() {
+		err1 := s.Start()
+		c.Assert(err1, IsNil)
+	}()
+	for i := 0; i < 10; i++ {
+		go func() {
+			fmt.Println("111")
+			s.Close()
```

Comment:
Close for multiple time is meaningless. You can use a loop to

for ... {
  waitGroup.Add(2)
  go func() { s.Start(); waitGroup.Done() }
  go func() { s.Close(); waitGroup.Done() }
  waitGroup.Wait()
}